[threaded-animation-resolution] don't accelerate size-dependent transform interpolation with if width and `height` are animated

graouts opened the pull request:
#### 126e9c6fc2313c7e6232a538f382b8dbb6f3224c
<pre>
[threaded-animation-resolution] don&apos;t accelerate size-dependent transform interpolation with if width and `height` are animated
<a href="https://bugs.webkit.org/show_bug.cgi?id=269563">https://bugs.webkit.org/show_bug.cgi?id=269563</a>

Reviewed by Dean Jackson.

We fixed a case in 272022@main where animating `width` or `height` while also animating `translate` or `transform: translate()`
with relative values would require transform interpolation to not be accelerated.

We now do the same work with threaded animation resolution, which makes the following WPT tests pass:

- css/css-transforms/animation/transform-percent-with-width-and-height-separate.html: Added.
- css/css-transforms/animation/transform-percent-with-width-and-height.html: Added.
- css/css-transforms/animation/translate-percent-with-width-and-height-separate.html: Added.
- css/css-transforms/animation/translate-percent-with-width-and-height.html: Added.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::canBeAccelerated const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):

Canonical link: <a href="https://commits.webkit.org/274815@main">https://commits.webkit.org/274815@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b191f0c07539c6eaf2a30576ac03ae5e8a161c4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40134 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19146 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42679 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36223 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42441 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22064 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16475 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40708 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/16122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/43957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/43957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12260 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/43957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5294 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -4068,9 +4068,17 @@ bool RenderLayerBacking::updateAcceleratedEffectsAndBaseValues()
+                auto& blendingKeyframes = effect->blendingKeyframes();
+                if ((blendingKeyframes.hasWidthDependentTransform() && animatesWidth)
+                    || (blendingKeyframes.hasHeightDependentTransform() && animatesHeight))
+                    continue;
```
Probably clearer as:

```
auto& blendingKeyframes = effect->blendingKeyframes();
if (animatesWidth && blendingKeyframes.hasWidthDependentTransform())
    continue;               
if (animatesHeight && blendingKeyframes.hasHeightDependentTransform())
    continue;               
```

graouts committed:
[threaded-animation-resolution] don't accelerate size-dependent transform interpolation with if width and `height` are animated
https://bugs.webkit.org/show_bug.cgi?id=269563

Reviewed by Dean Jackson.

We fixed a case in 272022@main where animating `width` or `height` while also animating `translate` or `transform: translate()`
with relative values would require transform interpolation to not be accelerated.

We now do the same work with threaded animation resolution, which makes the following WPT tests pass:

- css/css-transforms/animation/transform-percent-with-width-and-height-separate.html: Added.
- css/css-transforms/animation/transform-percent-with-width-and-height.html: Added.
- css/css-transforms/animation/translate-percent-with-width-and-height-separate.html: Added.
- css/css-transforms/animation/translate-percent-with-width-and-height.html: Added.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::canBeAccelerated const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAcceleratedEffectsAndBaseValues):

Canonical link: https://commits.webkit.org/274815@main
